subsys/usb/usb_c/usbc_pe_src_states.c

```diff
@@ -0,0 +1,691 @@
+	 *
+	 * NOTE: 1, 2 and VCONN off are done by Device Policy Manager when
+	 *	 it receives the HARD_RESET_RECEIVED notification.
+	 */
+	policy_notify(dev, HARD_RESET_RECEIVED);
+	pe->data_role = TC_ROLE_DFP;
+	policy_notify(dev, DATA_ROLE_IS_DFP);
+}
+
+void pe_src_transition_to_default_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Startup
+	 * state when:
+	 *	1: The Device Policy Manager indicates that the power
+	 *	   supply has reached the default level.
+	 */
+	if (policy_check(dev, CHECK_SRC_PS_AT_DEFAULT_LEVEL)) {
+		pe_set_state(dev, PE_SRC_STARTUP);
+	}
+}
+
+void pe_src_transition_to_default_exit(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * On exit from the PE_SRC_Transition_to_default state the
+	 * Policy Engine Shall:
+	 *	1: request the Device Policy Manager to turn on VCONN
+	 *	2: inform the Protocol Layer that the Hard Reset is complete.
+	 *
+	 * NOTE: The Device Policy Manager turns on VCONN when it notifies the
+	 *	 PE that the Power Supply is at the default level.
+	 */
+	prl_hard_reset_complete(dev);
+}
+
+/**
+ * 8.3.3.2.8 PE_SRC_Capability_Response State
+ */
+void pe_src_capability_response_entry(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * On entry to the PE_SRC_Capability_Response state the Policy Engine
+	 * Shall request the Protocol Layer to send one of the following:
+	 */
+
+	/*
+	 * 1: Reject Message – if the request cannot be met or the present
+	 *    Contract is Invalid.
+	 */
+	if (pe->snk_request_reply == SNK_REQUEST_REJECT) {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_REJECT);
+	}
+	/*
+	 * 2: Wait Message – if the request could be met later from the Power
+	 *    Reserve. A Wait Message Shall Not be sent if the present Contract
+	 *    is Invalid.
+	 */
+	else {
+		pe_send_ctrl_msg(dev, PD_PACKET_SOP, PD_CTRL_WAIT);
+	}
+}
+
+void pe_src_capability_response_run(void *obj)
+{
+	struct policy_engine *pe = (struct policy_engine *)obj;
+	const struct device *dev = pe->dev;
+
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Ready state when:
+	 *	1: There is an Explicit Contract AND
+	 *	2: A Reject Message has been sent and the present Contract
+	 *	   is still Valid OR
+	 *	3: A Wait Message has been sent.
+	 */
+	if (atomic_test_bit(pe->flags, PE_FLAGS_EXPLICIT_CONTRACT) &&
+			((pe->snk_request_reply == SNK_REQUEST_REJECT &&
+			policy_check_present_contract(dev, pe->present_contract)) ||
+			(pe->snk_request_reply == SNK_REQUEST_WAIT))) {
+		pe_set_state(dev, PE_SRC_READY);
+	}
+	/*
+	 * The Policy Engine Shall transition to the PE_SRC_Hard_Reset state
+	 * when:
+	 *	1: There is an Explicit Contract and
+	 *	2: The Reject Message has been sent and the present Contract
```

Comment:
For the requirement that the "reject message has been sent", do we need to check TX complete?

Reply:
I added the check